### PR TITLE
[FLINK-24597][state] fix the problem that RocksStateKeysAndNamespaceIterator would return duplicate data

### PR DESCRIPTION
**Pull request by mayuehappy**
…terator would return duplicate data

## What is the purpose of the change

fix the problem that RocksStateKeysAndNamespaceIterator would return duplicate data

## Brief change log

- Added the logic of filtering the same keyAndNamespace to the next method of RocksStateKeysAndNamespaceIterator
- add unit test for the case that with MapState and update userKey multiple times under the same key

## Verifying this change

Added test that validates that duplicate data could be filtered in `testIteratorUpdateMapStateMultipleTimes`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): ( no /)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: ( no )
  - The S3 file system connector: ( no )

## Documentation

  - Does this pull request introduce a new feature?  no)

### CODE REVIEW

**Comment by mayuehappy** [comment]
@flinkbot run azure

**Comment by mayuehappy** [comment]
@flinkbot run azure

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBRocksStateKeysAndNamespacesIteratorTest.java [hunk 1]
```diff
@@ -129,4 +131,74 @@ public void testIterator() throws Exception {
             }
         }
     }
+
+    @Test
+    public void testIteratorUpdateMapStateMultipleTimes() throws Exception {
```
Comment (Myasuka) [inline flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBRocksStateKeysAndNamespacesIteratorTest.java:136]: First of all, this test has many code same as `#testIteratorHelper`. Secondly, I think we should add new test in `StateBackendTestBase` to verify all state-backends.

**Review by Myasuka — CHANGES_REQUESTED** [review]
Thanks for your contribution! Please take a look at my comments.

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBRocksStateKeysAndNamespacesIteratorTest.java (hunk 1 above)
Comment (mayuehappy) [reply flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBRocksStateKeysAndNamespacesIteratorTest.java]: Very Thanks for reviewing. I total agree with your opinion and I've moved the test to `StateBackendTestBase`.
BTW , I think it is unnecessary to add a new test but just need to add some MapState-related case in `#testGetKeysAndNamespaces`. what do you think ?

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java [hunk 2]
```diff
@@ -460,22 +461,66 @@ public void testGetKeysAndNamespaces() throws Exception {
                     backend.getPartitionedState(
                             ns1,
                             StringSerializer.INSTANCE,
-                            new ValueStateDescriptor<>(fieldName, IntSerializer.INSTANCE));
+                            new ValueStateDescriptor<>(
+                                    valueStateFieldName, IntSerializer.INSTANCE));
+
+            MapState<Integer, Integer> keyedState2 =
```
Comment (Myasuka) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:467]: I think modify existing test is okay, however, I don't we should introduce so much code duplication.
I think you could leverage `backend.createInternalState` API which could set current namespace of `ns1` or `ns2`.

**Review by Myasuka — CHANGES_REQUESTED** [review]
Thanks for the update, please consider to refactor current test.

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java (hunk 2 above)
Comment (mayuehappy) [reply flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:467]: OK , I noticed that there is a pair of unit tests `#testGetKeys` and `#testMapStateGetKeys` in `StateBackendTestBase`, which is very similar to my situation. I referred to these codes when I modified it before.

Comment (mayuehappy) [reply flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:467]: In addition, I observed that similar tests in StateBackendTestBase seem to have some duplicate codes.(eg `testListStateRestoreWithWrongSerializers`\ `testValueStateRestoreWithWrongSerializers` `testMapStateRestoreWithWrongSerializers`)
So maybe it's also fine to add a new test as `#testMapStateGetKeysAndNamespace` just like` #testMapStateGetKeys ` or do u have some other greater suggestions ?

Comment (Myasuka) [reply flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:467]: I think you could add a new `#testMapStateGetKeysAndNamespace` method.

Comment (mayuehappy) [reply flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java]: Thanks for your advice, have resolved it .:)

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java [hunk 3]
```diff
@@ -5084,6 +5084,58 @@ public void testMapStateGetKeys() throws Exception {
...
+                keyedState1.put("ho", key * 2);
+                keyedState2.put("he", key * 2);
+                keyedState2.put("ho", key * 2);
+            }
+
+            try (Stream<Tuple2<Integer, String>> stream = backend.getKeysAndNamespaces(fieldName)) {
+                final Map<String, Set<Integer>> keysByNamespace = new HashMap<>();
+                stream.forEach(
+                        entry -> {
+                            assertThat("Unexpected namespace", entry.f1, isOneOf(ns1, ns2));
```
Comment (Myasuka) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:5122]: As I suggested before, you could leverage `createInternalState` API to reduce code duplication.

I think code below looks better:
~~~ java
InternalMapState<Integer, String, String, Integer> internalState = backend.createInternalState(
        StringSerializer.INSTANCE,
        new MapStateDescriptor<>(
                fieldName, StringSerializer.INSTANCE, IntSerializer.INSTANCE));
String[] namespaces = new String[] {"ns1", "ns2"};

for (int key = 0; key < elementsNum; key++) {
    backend.setCurrentKey(key);
    for (String ns : namespaces) {
        internalState.setCurrentNamespace(ns);
        internalState.put("hello", key);
        internalState.put("world", key);
    }
}

try ...........
 assertThat("Unexpected namespace", entry.f1, isOneOf(namespaces)
~~~

**Review by Myasuka — CHANGES_REQUESTED** [review]
Thanks for the update, please take a look at my comments.

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java (hunk 3 above)
Comment (mayuehappy) [reply flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java]: Thank you very much for your patience, sorry for not understanding your meaning before. It seems that the code will be more concise. I have updated it.

**Review by Myasuka — APPROVED** [review]
LGTM and merged.

**Pull request closed without merging**